### PR TITLE
catalog-import: stop trying to register optional locations

### PR DESCRIPTION
**Pull request by Rugvip**
See #9663, as far as I can tell it's ignored by the backend. Am I missing some point where it would be handled? 😅

### CODE REVIEW

**Comment by freben** [comment]
As far as I can see we do honor this. https://github.com/backstage/backstage/blob/8651981680e76a8b1198075102d2dbc3ea9fbdc6/plugins/catalog-backend/src/providers/DefaultLocationStore.ts#L62

**Comment by Rugvip** [comment]
@freben no presence here tho right? Which is what's being converted there https://github.com/backstage/backstage/blob/8651981680e76a8b1198075102d2dbc3ea9fbdc6/plugins/catalog-backend/src/providers/DefaultLocationStore.ts#L52

**Comment by freben** [comment]
@Rugvip Yeah but isn't it ultimately the generated `kind: Location` entity that matters?

**Comment by Rugvip** [comment]
@freben sure, but the presence isn't forwarded there as far as I can see.

**Comment by freben** [comment]
But then maybe _that_ is the problem? 🤔 Isn't it?

**Comment by pjungermann** [comment]
not sure if this is related, but thought its better to mention it here: At least some of the catalog discoveries will register Location entities with presence option. E.g., the bitbucket-discovery creates a optional Location for each repository, at that point not being aware whether the catalog-info.yaml from the registered Location exists or not.

**Comment by Rugvip** [comment]
@pjungermann yep, totally. This doesn't have an effect on that flow though, it's just for registered locations

**Comment by webark** [comment]
>  the bitbucket-discovery creates a optional Location for each repository

Is this the desired behavoir? Or just a current artifact of the bitbucket integration.

**Comment by pjungermann** [comment]
> > the bitbucket-discovery creates a optional Location for each repository
> 
> Is this the desired behavoir? Or just a current artifact of the bitbucket integration.

I think there is no real alternative for this so far. If you don't want to add optional Location entities, you need to know what exists. How it works though is that repositories will be discovered - without knowing whether they have a catalog file or not. Hence, all of them will be registered and then scanned by the update loop going through those Locations and checking their output. Of course, the same check could be done as part of the processor, but that would require a call per potential Location/repository. With potential trouble with rate limiting, etc. (which you may have later, too, though).

Basically, it postpones this check to the update loop.

This might be due to the limits of the Bitbucket Cloud/Server APIs, but is likely a problem which does not exist only for this discovery.

I proposed an idea for improvement for the bitbucket discovery at issue #9765 with a pending PR.

**Comment by webark** [comment]
@pjungermann I had a random thought today about using an entity provider that taps into the bitbucket search and generates a location (or maybe a location per project) that generates a list of targets from the search results. 

Is this a bad idea? Would it work?

**Comment by pjungermann** [comment]
@webark This is basically what I did with my PR #9924 for issue #9765. Just that it is still "just" a processor and not a provider.

@freben mentioned at a comment there, too, that it could be a good candidate for a provider.

Generally, I agree. I assume that for a provider we would need to actually fetch the catalog files in order to register the contained entities, considering updates, etc.

This might be a bit tricky to realize even though I added ETag handling at PR #10043 due to the fact that ETags are broken at Bitbucket Cloud's API endpoint to fetch the file content..

But if I got @Rugvip'c reply to my question above right, the behavior to register optional locations at the discovery would not impact the scope here. Also, other discovery processors behave similarly.

**Comment by Rugvip** [comment]
Discussed this a bit more with @freben and we've realized what's actually going on here. Back when we migrated to the new catalog we inadvertently changed the behavior of the `POST /locations` endpoint. The big change was that the `dryRun` flag became a completely separate codepath, and the only codepath that actually verified that the location was processable.

In the old catalog we used to try to process the location regardless, which is why this flag was added in the first place, so that you could register locations that didn't exist yet. The way you do that nowadays is that you simply skip the dry run step and just add the location directly, which the catalog will allow.

All in all it means we do indeed want to remove this flag as it's redundant, and it will have no impact on the discovery flows as that is handled separately.

**Review by freben — APPROVED** [review]
